src/lib/features/client-feature-toggles/delta/client-feature-toggle-delta.ts

```diff
@@ -18,92 +18,75 @@ import { CLIENT_DELTA_MEMORY } from '../../../metric-events';
 import type EventEmitter from 'events';
 import type { Logger } from '../../../logger';
 import type { ClientFeaturesDeltaSchema } from '../../../openapi';
-
-type DeletedFeature = {
-    name: string;
-    project: string;
-};
-
-export type RevisionDeltaEntry = {
-    updated: FeatureConfigurationDeltaClient[];
-    revisionId: number;
-    removed: DeletedFeature[];
-    segments: IClientSegment[];
-};
-
-export type Revision = {
-    revisionId: number;
-    updated: any[];
-    removed: DeletedFeature[];
-};
-
-type Revisions = Record<string, RevisionDelta>;
-
-const applyRevision = (first: Revision, last: Revision): Revision => {
-    const updatedMap = new Map(
-        [...first.updated, ...last.updated].map((feature) => [
-            feature.name,
-            feature,
-        ]),
-    );
-    const removedMap = new Map(
-        [...first.removed, ...last.removed].map((feature) => [
-            feature.name,
-            feature,
-        ]),
+import {
+    DELTA_EVENT_TYPES,
+    type DeltaEvent,
+    type DeltaHydrationEvent,
+    isDeltaFeatureRemovedEvent,
+    isDeltaFeatureUpdatedEvent,
+} from './client-feature-toggle-delta-types';
+
+type EnvironmentRevisions = Record<string, DeltaCache>;
+
+export const filterEventsByQuery = (
+    events: DeltaEvent[],
+    requiredRevisionId: number,
+    projects: string[],
+    namePrefix: string,
+) => {
+    const targetedEvents = events.filter(
+        (revision) => revision.eventId > requiredRevisionId,
     );
-
-    for (const feature of last.updated) {
-        removedMap.delete(feature.name);
-    }
-
-    for (const feature of last.removed) {
-        updatedMap.delete(feature.name);
-        removedMap.set(feature.name, feature);
-    }
-
-    return {
-        revisionId: last.revisionId,
-        updated: Array.from(updatedMap.values()),
-        removed: Array.from(removedMap.values()),
+    const allProjects = projects.includes('*');
+    const startsWithPrefix = (revision: DeltaEvent) => {
+        return (
+            (isDeltaFeatureUpdatedEvent(revision) &&
+                revision.feature.name.startsWith(namePrefix)) ||
+            (isDeltaFeatureRemovedEvent(revision) &&
+                revision.featureName.startsWith(namePrefix))
```

Comment:
Why are these different?

Reply:
Feature updated has complex object of feature, with all the description, strategies, constraints. Removed has just string value.